src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,129 @@
+using Amazon.S3;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress.S3
+{
+    /// <summary>s
+    /// Egress provider for egressing stream data to the S3 storage.
+    /// </summary>
+    internal class S3StorageEgressProvider : EgressProvider<S3StorageEgressProviderOptions>
+    {
+        internal class StorageFactory
+        {
+            public virtual async Task<IS3Storage> CreateAsync(S3StorageEgressProviderOptions options, EgressArtifactSettings settings, CancellationToken cancellationToken) => await S3Storage.CreateAsync(options, settings, cancellationToken);
+        }
+
+        internal StorageFactory ClientFactory = new();
+
+        public S3StorageEgressProvider(ILogger<S3StorageEgressProvider> logger) : base(logger)
+        {
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<CancellationToken, Task<Stream>> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await using var stream = await action(token);
+
+                var client = await ClientFactory.CreateAsync(options, artifactSettings, token);
+                if (stream.CanSeek) // use the stream directly
+                {                    
+                    await client.PutAsync(stream, token);
+                }
+                else // copy temporary to memory stream locally
+                {
+                    await client.UploadAsync(stream, token);
+                }
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (AmazonS3Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<Stream, CancellationToken, Task> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            IS3Storage client = null;
+            string uploadId = null;
+            bool uploadDone = false;
+            try
+            {
+                client = await ClientFactory.CreateAsync(options, artifactSettings, token);
+                uploadId = await client.InitMultiPartUploadAsync(artifactSettings.Metadata, token);
+                await using var stream = new MultiPartUploadStream(client, options.BucketName, artifactSettings.Name, uploadId, options.CopyBufferSize!.Value);
```

Comment:
I don't see where `CopyBufferSize` has a default value set, so this is going to be `null` and throw an NRE for the default usage.